### PR TITLE
Add player

### PR DESCRIPTION
**Pull request by lrharris215**
- The game now has a player and a board class. The user can pick 1 position to place their mark, and the board will update to show their mark in that position.

### CODE REVIEW

File: lib/game.rb [hunk 1]
```diff
@@ -1,23 +1,40 @@
...
-  attr_reader :display
-  def initialize(display, presenter)
+  attr_reader :display, :board
+  def initialize(display, presenter, board)
     @display = display
     @presenter = presenter
+    @board = board
+    @player_1 = Player.new("X")
+    @player_2 = Player.new("O")
+    @game_over = false
```
Comment (kudukacar) [inline lib/game.rb:13]: Why do we need two players? Where do we use @game_over?

File: spec/game_spec.rb [hunk 2]
```diff
@@ -13,27 +13,64 @@ def output(message)
...
+end
+
 describe "Game" do
+  before(:each) do
+    @game = Game.new(TestDisplay.new, TestPresenter.new, TestBoard.new)
+  end
   it "passes this example test" do
     pass = true
     expect(pass).to eq(true)
   end
```
[the remark this answers is not in the document]
Comment (lrharris215) [reply spec/game_spec.rb:50]: Ok, wasn't sure when I was supposed to do that.

File: lib/game.rb [hunk 3]
```diff
@@ -1,23 +1,40 @@
...
-game.play
+  def place_mark(player, position)
+    pos = position.to_i
+    @board[pos - 1] = player.mark
+  end
+
+  def get_player_input
+    @display.output("Where would you like to put your mark?")
+    gets.chomp
+  end
```
[the remark this answers is not in the document]
Comment (lrharris215) [reply lib/game.rb:39]: - Will work on abstracting gets, stdin 
- Yeah I wasn't sure where to put the responsibility of getting input, I can try reworking it so game has less to do
- Validations are coming in a separate PR. I didn't want to get too ahead of myself on what was/wasn't allowed to be entered.

File: spec/game_spec.rb [hunk 4]
```diff
@@ -13,27 +13,64 @@ def output(message)
...
+class TestPlayer
+  def mark
+    "Z"
+  end
+end
+
 describe "Game" do
+  before(:each) do
+    @game = Game.new(TestDisplay.new, TestPresenter.new, TestBoard.new)
+  end
```
[the remark this answers is not in the document]
Comment (lrharris215) [reply spec/game_spec.rb:45]: For some reason it didn't work when the var was just game.

File: lib/game.rb (hunk 1 above)
Comment (lrharris215) [reply lib/game.rb:13]: This is where I started getting ahead of myself. I commented this out for now.

File: lib/board.rb [hunk 5]
```diff
@@ -0,0 +1,19 @@
...
+  end
+
+  def []=(num, val)
+    @board[num] = val
+  end
+
+  def [](num)
+    @board[num]
+  end
+
```
[the remark this answers is not in the document]
Comment (lrharris215) [reply lib/board.rb:15]: So this one is just so that the presenter can index into the board. I suppose I could have just passed in board.board?

File: lib/board.rb [hunk 6]
```diff
@@ -0,0 +1,19 @@
...
+  attr_reader :board
+
+  def initialize
+    @board = [1, 2, 3, 4, 5, 6, 7, 8, 9]
+  end
+
+  def []=(num, val)
+    @board[num] = val
+  end
+
```
[the remark this answers is not in the document]
Comment (lrharris215) [reply lib/board.rb:11]: You're right, that was an oversight. Deleted this now, don't need it anymore.

File: lib/board.rb [hunk 7]
```diff
@@ -0,0 +1,16 @@
...
+  attr_reader :board
+
+  def initialize
+    @board = [1, 2, 3, 4, 5, 6, 7, 8, 9]
+  end
+
+  def [](num)
+    @board[num]
+  end
+
```
Comment (kudukacar) [inline lib/board.rb:11]: Just like you have place_mark below, can we have a more descriptive name for this method?  For example, @board.place_mark tells me exactly what it's doing.  board[0] exposes the data structure and in plain english does not tell me what it's doing.

File: lib/displayer.rb [hunk 8]
```diff
@@ -1,9 +1,14 @@
...
   end
 
   def output(message)
     @output_function.puts(message)
   end
+
+  def input
+    @input_function.gets.chomp
+  end
 end
```
Comment (kudukacar) [inline lib/displayer.rb:14]: Awesome!

File: spec/game_spec.rb [hunk 9]
```diff
@@ -13,27 +13,53 @@ def output(message)
 end
 
 class TestPresenter
-  def print_board
-    "______"
+  def print_board(board)
+    board.board
   end
```
Comment (kudukacar) [inline spec/game_spec.rb:18]: board.board sounds odd.  Is there a better name than board for the instance variable? I'm not good with naming.

File: spec/game_spec.rb (hunk 4 above)
[the remark this answers is not in the document]
Comment (kudukacar) [reply spec/game_spec.rb:45]: Perhaps you can set the subject of game at the beginning of the test, but this is fine as well

File: spec/game_spec.rb [hunk 10]
```diff
@@ -13,27 +13,53 @@ def output(message)
...
+    expect(@game.display.state[0]).to eq([1, 2, 3])
+  end
+
+  it "Allows a player to place their mark on the board" do
+    player = TestPlayer.new
+
+    @game.board.place_mark(player, 1)
+    @game.print_board
+    expect(@game.display.state[0][0]).to eq("Z")
   end
```
Comment (kudukacar) [inline spec/game_spec.rb:64]: This is an interesting test in that you don't actually call @game.play but test a couple of methods within play.  Do you think this is a good way to test?  We want to test our public facing methods, for example @game.play.  @game.print_board should be a private method, as it's really never used outside of game.  Really the second test for game is it "plays the game" and you could show the sequence of boards displayed.  
But this hits on the problem with "newing up" a model in initialize - it's very difficult to unit test.  If you passed in an instance of player to game, as you do with display, presenter, and board, then you could easily test the play method.

File: lib/player.rb [hunk 11]
```diff
@@ -0,0 +1,12 @@
+class Player
+  attr_reader :mark
+  def initialize(mark)
+    # X or O
+    @mark = mark
+  end
+
+  def get_input(display)
```
Comment (pumpkincouture) [inline lib/player.rb:8]: When we think about display (things logged to the console), can we place them in a class to encapsulate that particular responsibility?

Comment (pumpkincouture) [reply lib/player.rb:8]: Ah, I see what you're doing, ie passing the display in places. Maybe instead of magic strings in various files, you can group them in one constants messages file and then access them when needed.

File: spec/game_spec.rb (hunk 9 above)
Comment (pumpkincouture) [reply spec/game_spec.rb:18]: Perhaps `spaces` or `cells`?

File: lib/board.rb (hunk 7 above)
Comment (lrharris215) [reply lib/board.rb:11]: Ok, renamed this to find_position

File: spec/game_spec.rb (hunk 9 above)
Comment (lrharris215) [reply spec/game_spec.rb:18]: I like the sound of cells. Renamed.

File: spec/game_spec.rb (hunk 10 above)
Comment (lrharris215) [reply spec/game_spec.rb:64]: Will have to think about this one, it's the largest change to implement. I get what you mean about the unit testing, I was getting worried that game was starting to take in too many arguments.

Comment (kudukacar) [reply spec/game_spec.rb:64]: It's okay to initialize the game with a player as you do with display, presenter, and board.  When you have two players, you could find a way to limit the arguments (for example, game could be initialized with an array of players).

Comment (lrharris215) [reply spec/game_spec.rb:65]: I started writing out a test for game.play, but then stopped bc I wasn't sure what the test should be. This isn't the final version of game.play, and I'd have to alter the test to accommodate a second player anyway. That's why I wrote tests for the functions within game.play, because those smaller functions are still going to be used later. 

The test I wrote ended up failing, and I'm still not entirely sure why either. It was pushing [X,2,3] into the display.state before and after the get_input prompt.

**Review by kudukacar — APPROVED** [review]
